[Fix #2719] App does not fit on iPhone X screen

foopang opened the pull request:
fixes #2719 

I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂

<img width="391" alt="screen shot 2017-12-30 at 9 42 17 pm" src="https://user-images.githubusercontent.com/4666757/34454682-5392b97e-edab-11e7-9f80-2c249267698b.png">

<img width="391" alt="screen shot 2017-12-30 at 9 39 35 pm" src="https://user-images.githubusercontent.com/4666757/34454683-578dfe08-edab-11e7-8662-1432238fc7f2.png">


status: ready <!-- Can be ready or wip -->

foopang committed:
[Fix #2719] App does not fit on iPhone X screen

Signed-off-by: Andrey Shovkoplyas <motor4ik@gmail.com>

denis-sharypin commented:
Hey, @foopang thanks for a contribution. Could you, please double check tab bar and top bar heights? Use my design as a reference. 
<img width="642" alt="iphonex" src="https://user-images.githubusercontent.com/6559881/34488205-ca67ee38-efd7-11e7-825a-23dc873fc932.png">

foopang commented:
Hi @denis-sharypin please see if this looks good now.

<img width="461" alt="screen shot 2018-01-03 at 6 29 25 am" src="https://user-images.githubusercontent.com/4666757/34502979-1f687e56-f051-11e7-92b7-97851f831c15.png">

foopang commented on src/status_im/ui/screens/views.cljs:
```diff
@@ -113,9 +113,25 @@
+                                                            :choose-recipient) common-styles/color-blue4
+                                                           (:accounts :login) common-styles/color-blue2
+                                                           :transparent)
+                                                         :margin-top -20}))
```
Seems that the new `safe-area-view` RN component doesn't allow us to set the padding size for top and bottom, so I come up with this kinda odd `:margin-top -20` to do the trick.

denis-sharypin commented:
@foopang  Thanks! All good from a design side

denis-sharypin commented:
@foopang Please, add also a launch screen 
![launchscreen1125x2436](https://user-images.githubusercontent.com/6559881/34514035-96b9bc4e-f06b-11e7-8cf0-6c1290c80fe5.png)

foopang commented:
@denis-sharypin the launch screen has been updated!

<img width="461" alt="screen shot 2018-01-03 at 8 37 05 pm" src="https://user-images.githubusercontent.com/4666757/34522335-17054128-f0cd-11e7-8a11-f91604e379ca.png">

oskarth commented:
@foopang Sorry, got dropped out of our pipeline. Moving to REVIEW now as design seems to have been sorted. Thanks!

oskarth commented on src/status_im/ui/screens/views.cljs:
```diff
@@ -113,9 +113,25 @@
+                                 (and (= width 375) (= height 812))))]
           [(if android? menu-context view) common-styles/flex
-           [view common-styles/flex
+           [(if iphone-x? safe-area-view view) (merge common-styles/flex
```
This is a bit gnarly, can we break this out in a fn or so?

oskarth commented on src/status_im/ui/screens/views.cljs:
```diff
@@ -1,8 +1,8 @@
-            [status-im.utils.platform :refer [android?]]
-            [status-im.ui.components.react :refer [view modal]]
+            [status-im.utils.platform :refer [android? ios?]]
+            [status-im.ui.components.react :refer [view safe-area-view modal dimensions]]
```
Can we turn these into react/foo etc? Generally preferable (think these are leftovers)

oskarth commented on src/status_im/ui/components/react.cljs:
```diff
@@ -36,6 +36,7 @@
 (def geolocation (when (exists? js/window)
                    js/navigator.geolocation.))
 (def view (get-class "View"))
+(def safe-area-view (get-class "SafeAreaView"))
```
Is this always available? What happens if it is not?

foopang replied:
Yes, it is always available. React Native ships with `SafeAreaView` as of version 0.50.

oskarth reviewed (changes requested):
Minor comments, in general looks good

oskarth commented:
^ @jeluard @flexsurfer please review

jeluard commented on src/status_im/ui/screens/views.cljs:
```diff
@@ -113,9 +113,25 @@
+                          (throw (str "Unknown view: " current-view)))
+              iphone-x?   (and ios?
+                               (let [{:keys [width height]} (-> (.get dimensions "window")
+                                                                (js->clj :keywordize-keys true))]
```
Sounds very error prone. Is that the official to do this?

jeluard reviewed (changes requested):
Seems very hacky. Can you provide link to docs detailing official guidelines?

foopang commented:
@jeluard I am not sure if there are official guidelines dealing with iPhone X from React Native docs. I followed this as a reference https://blog.callstack.io/react-native-meets-iphonex-a0f5034f7a41

jeluard commented on src/status_im/ui/components/react.cljs:
```diff
@@ -202,3 +203,25 @@
+
+(defn create-main-screen-view [current-view]
+  (let [iphone-x?    (and ios?
+                          (let [{:keys [width height]} (-> (.get dimensions "window")
```
Can we extract this logic somewhere?

jeluard commented on src/status_im/ui/components/react.cljs:
```diff
@@ -202,3 +203,25 @@
+                          (let [{:keys [width height]} (-> (.get dimensions "window")
+                                                           (js->clj :keywordize-keys true))]
+                            ;; iPhone X dimensions = 375 x 812
+                            (and (= width 375) (= height 812))))
```
Please use constant for magic number

foopang commented:
@jeluard updated, please have a look.

jeluard commented:
@foopang Thanks!

Serhy commented:
@foopang there is conflict appeared, could you rebase with develop please? 

Additionally, we tested on iPhoneX (iOS 11.2.2) the build just right before the conflict -  https://i.diawi.com/QNR1AK (0.0.2157 version).

Adjusting the views to iPhoneX we introduce several issues we need to consider before merge:

**1)** On QR code screens (both - wallet and public key). Could you please move the button with the key value and QR code a bit higher so the button not overlapped by home indicator.
<img width="365" alt="screen shot 2018-01-25 at 18 28 18" src="https://user-images.githubusercontent.com/8749671/35433593-b4ecf338-028c-11e8-8cbc-526a5e3303d2.png">


**2)** The **Contacts** view (_Wallet -> Send transaction -> Choose recipient… -> Choose from contacts_) and **Transaction Fee** view headers (_Wallet -> Send transaction -> Advanced -> Transaction Fee_) are too close to device status icons menu. Could you move both screens a bit lower please? (Green line on Transaction Fee screen just to show the edge of the device status icons menu)
<img width="368" alt="screen shot 2018-01-25 at 18 31 47" src="https://user-images.githubusercontent.com/8749671/35434593-6313c4b6-0290-11e8-92bd-1c3b18f43b2d.png">
<img width="367" alt="screen shot 2018-01-25 at 18 32 10" src="https://user-images.githubusercontent.com/8749671/35434594-6352249a-0290-11e8-9490-668243e2c953.png">

**3)** There is a gap between keyboard and message editbox appears when in 1-1/group/public chat user types a message. Could you move please messagebox area lower and join it with the keyboard?
<img width="365" alt="screen shot 2018-01-25 at 18 33 01" src="https://user-images.githubusercontent.com/8749671/35434702-b7ea0324-0290-11e8-9a65-19cea3796848.png">

**4)** Pop-up screen menu with the list of contacts from /request option in Group Chat overlapped by device status icons menu. (to reproduce - in any Group Chat select /request option and pull-top the appeared menu with the list of contacts)
<img width="366" alt="screen shot 2018-01-25 at 18 36 53" src="https://user-images.githubusercontent.com/8749671/35434756-e821df08-0290-11e8-80e6-61832332f72b.png">

5) The @browse screen extended pretty much so it’s overlapped by device status icons menu. **To reproduce**: Open Status -> Create new account -> create password -> in Console tap on @browse -> type amazon.com -> tap Send button -> **Result**: the first screenshot below -> select 'Switch to fullscreen' at the bottom-right -> **Result**: the second screenshot below). 
<img width="367" alt="screen shot 2018-01-25 at 18 41 50" src="https://user-images.githubusercontent.com/8749671/35434897-6f155cce-0291-11e8-808e-0c023698bfe0.png">
<img width="366" alt="screen shot 2018-01-25 at 18 42 08" src="https://user-images.githubusercontent.com/8749671/35434898-6f45b36a-0291-11e8-94c5-cb3b180f3bf9.png">

6) The **Assets** (_Wallet -> top-right menu button -> Manage Assets_) and **Filter history** (_Wallet -> Transaction history -> Top right menu button_) screens overlapped by device status icons menu. It is not possible to tap on ‘Done’ button to return to previous screen. 
<img width="364" alt="screen shot 2018-01-25 at 18 44 02" src="https://user-images.githubusercontent.com/8749671/35435087-1a08a26c-0292-11e8-90a7-0918abb22c4d.png">
<img width="365" alt="screen shot 2018-01-25 at 18 44 40" src="https://user-images.githubusercontent.com/8749671/35435089-1a23e68a-0292-11e8-9eb3-5b6d90f6fd85.png">

7) The tab bar on the Wallet main screen is cut off, - not the same as on Profile or Home screens. Could you extend please the white background to the bottom of the tab bar on Wallet screen view.
<img width="365" alt="screen shot 2018-01-25 at 18 49 46" src="https://user-images.githubusercontent.com/8749671/35435170-61e4b7ba-0292-11e8-8ba7-a1d2c12093c9.png">

foopang commented:
I’ve refactored the code a bit in order to address the issues found by @Serhy 

For issue #5 reported, it is not always reproducible. I suspect that is a bug of `react-native-webview-bridge` but I couldn’t fix it yet.

<img width="461" alt="screen shot 2018-01-27 at 2 00 28 pm" src="https://user-images.githubusercontent.com/4666757/35472050-289550de-03a3-11e8-8de7-367c5eb0524a.png">
<img width="461" alt="screen shot 2018-01-27 at 2 01 08 pm" src="https://user-images.githubusercontent.com/4666757/35472051-28c559a0-03a3-11e8-9f43-8337f882f63c.png">
<img width="461" alt="screen shot 2018-01-27 at 2 08 40 pm" src="https://user-images.githubusercontent.com/4666757/35472052-28f459c6-03a3-11e8-94d0-033d49322766.png">
<img width="461" alt="screen shot 2018-01-27 at 2 47 36 pm" src="https://user-images.githubusercontent.com/4666757/35472053-2920d7c6-03a3-11e8-955c-1bc484c9bf0f.png">
<img width="461" alt="screen shot 2018-01-27 at 4 02 11 pm" src="https://user-images.githubusercontent.com/4666757/35472054-294c2e08-03a3-11e8-95d1-5ad69edc7b87.png">
<img width="461" alt="screen shot 2018-01-27 at 7 26 00 pm" src="https://user-images.githubusercontent.com/4666757/35472055-297748e0-03a3-11e8-9bb3-89067cf0e670.png">
<img width="461" alt="screen shot 2018-01-27 at 8 00 04 pm" src="https://user-images.githubusercontent.com/4666757/35472056-29a27998-03a3-11e8-8b3c-99a7b834eb1d.png">
<img width="461" alt="screen shot 2018-01-28 at 5 31 07 pm" src="https://user-images.githubusercontent.com/4666757/35481046-af76433e-0456-11e8-9eae-926b9aaeb938.png">

jeluard commented:
@foopang Have you considered handling this at the `status-bar` level instead of relying on `platform-specific-view`?

foopang commented:
@jeluard Oh, you're right! I can simplify it a little bit and also remove the odd looking `margin-top -20` in `platform-specific-view` by just setting `:status-bar-default-height` to 0 for iPhone X. Thanks for the useful tip!

foopang commented:
@Serhy I think the @browse screen issue is not relevant anymore since we removed the @browse command and implemented our own browser. Can you check this again? 

<img width="461" alt="screen shot 2018-02-04 at 11 22 35 am" src="https://user-images.githubusercontent.com/4666757/35777303-a653ed0c-09e6-11e8-9ea0-3a9c46d9ff9d.png">

Serhy commented:
Thanks a lot @foopang 
Build for iOS: https://i.diawi.com/oUg64n has been tested on iPhoneX.

Yes, all above issues are fixed, and with the removed @browse screen - 5th issue from https://github.com/status-im/status-react/pull/2856#issuecomment-360741904 is not relevant anymore. A couple of things appeared we need to look at (I could not reproduce them in latest develop which is 0.9.11d632):

On `Wallet` -> `Send transaction` and `Wallet` -> `Receive transaction` screens there is bottom bar with buttons: 'Sign later'/'Sign transaction' (for 'Send transaction' screen) and 'Share'/'Send request' (for 'Receive transaction' screen). 
These bars apparently behind keyboard, when the last is present on screen, so there is no way to proceed with the request further until keyboard is hidden. Please see steps to reproduce what I mean:

1) Open Status and log in to account
2) Navigate to Wallet -> 'Send transaction'
3) 'Choose recipient' -> 'Choose from contacts' -> tap 'Jarrad'
4) Tap in 'Specify amount' 
5) Tap 'Sign later'/'Sign transaction'

**Actual result(1)**: 
![image uploaded from ios 2](https://user-images.githubusercontent.com/8749671/35818613-38ddab8c-0aa9-11e8-8d6a-3ee59cae9eeb.jpg)

6) Navigate back to Wallet main screen
7) Back to 'Send transaction' 
8) Tap Specify amount and type 0
9) Tap 'Choose recipient' -> Choose From Contacts -> Jarrad (Note: in such case the bottom bar with 'Sign later' and 'Sign transaction' buttons are visible, however:
10) Tap 'Sign transaction' enter password and look if 'Cancel' with 'Sign transaction' buttons are present

**Actual result(2)**: 
![image uploaded from ios 5](https://user-images.githubusercontent.com/8749671/35818779-92033362-0aa9-11e8-8a1e-8fea2f75de68.jpg)


11) Navigate to 'Wallet' -> 'Receive transaction' screen 
12) Tap 'Specify amount' and type 0
13) Tap 'Send request' 

**Actual request(3)**: 
![image uploaded from ios 3](https://user-images.githubusercontent.com/8749671/35818702-66a29d0c-0aa9-11e8-9e06-64e89d14d3e4.jpg)

And one more issue with 'Sign transaction' screen
1) From 'Home' screen -> tap `+` (Start new chat) -> Jarrad (to start 1-1 chat)
2) Tap /send command -> enter 0 value -> 'Send message' blue button
3) Once you wind up on 'Send transaction' screen check bottom bar with the 'Sign later' and 'Sign transaction' buttons

**Actual result(4)**: 
![image uploaded from ios 4](https://user-images.githubusercontent.com/8749671/35818974-17021a2e-0aaa-11e8-9cb8-87cee0619f18.jpg)


Could you move them upper a bit so it would not be overlapped by the device home indicator?

janherich commented:
@foopang Please resolve conflicts

foopang commented:
@Serhy The transaction screen issues have been fixed now. Please test again and let me know if there are any issues still. Thanks!

<img width="461" alt="screen shot 2018-02-06 at 11 12 55 am" src="https://user-images.githubusercontent.com/4666757/35852857-f1003334-0b66-11e8-963e-61a5ec07a3c6.png">
<img width="461" alt="screen shot 2018-02-06 at 11 17 40 am" src="https://user-images.githubusercontent.com/4666757/35852862-f411beee-0b66-11e8-929c-a285c010102b.png">
<img width="461" alt="screen shot 2018-02-06 at 11 19 18 am" src="https://user-images.githubusercontent.com/4666757/35852867-f6704e44-0b66-11e8-89b6-0e6154765a2d.png">
<img width="461" alt="screen shot 2018-02-06 at 3 43 27 pm" src="https://user-images.githubusercontent.com/4666757/35852876-fb47f692-0b66-11e8-83c4-a42bda763e66.png">

Serhy commented:
Hey @foopang, thanks! All above fixed in latest build for iOS: https://i.diawi.com/V2TuaJ.
But there is one regression appeared (might be related to fix of the last one issue where transaction is send and processed from chat):

There is no 'Transaction sent' confirmation screen after transaction sign in, if Tx was created from 1-1/group chats or signed from 'Transactions' -> 'Unsigned' screen. App is no responsive for tap on any part of screen until re-open the app. 
Please see steps to reproduce: 

**Preconditions**: Account is created and ETH balance > 0.0001 (/faucet request sent) 

1) Open Status and log in to account
2) Navigate to 'Home' -> 'Start new chat' -> 'Jarrad'
3) Tap on /send option -> type valid amount to send (e.g. 0.00003) -> 'Send' button
4) From 'Send transaction' ->tap  'Sign transaction' button -> enter valid password -> 'Sign transaction' button
5) Check if you wind up on 'Transaction sent' screen

Actual result: Redirect to 1-1 chat screen. App is not responsive. 
![image uploaded from ios 7](https://user-images.githubusercontent.com/8749671/35865266-bbb582b0-0b5c-11e8-829d-731390d12073.jpg)


6) Re-open Status application and log in with the same account
7) Navigate to 'Wallet' -> 'Send transaction'
8) Specify recipient -> Recent recipients -> Jarrad
9) Specify amount -> type valid amount (e.g. 0.00005)
10) Tap 'Sign later' -> Yes
11) Navigate to 'Transaction history' -> 'Unsigned' tab -> Tap 'Sign' button on unsigned transaction
12) From 'Send transaction' ->tap  'Sign transaction' button -> enter valid password -> 'Sign transaction' button 
13) Check if you wind up on 'Transaction sent' screen

Actual result: Redirect to 'Transactions' screen. App is not responsive. 
![image uploaded from ios 6](https://user-images.githubusercontent.com/8749671/35865259-b77e8796-0b5c-11e8-8120-5bae6e75c251.jpg)

Note: on both screens above - there is not visible device status bar at the very top.


And one more thing I have to note (after rebase from develop appeared):
The camera view screen which accessed from 'Wallet' -> 'Send transaction' -> 'Scan QR code': the there is a rectangle instead of square for QR scanning area and two black unexpected squares below are present at the moment:

<img width="638" alt="screen shot 2018-02-06 at 17 36 20" src="https://user-images.githubusercontent.com/8749671/35868111-487f38a6-0b64-11e8-9fbf-4acba6f6a96b.png">

foopang commented:
@Serhy Sorry for bringing in so many issues in this PR. Hope all is fixed now! Please have a look.

Serhy commented:
Thanks a lot, @foopang !
We checked https://i.diawi.com/nZrTp8 build on iPhoneX and iPhone 5s (and Android 7.0 too!) as well this PR and it was noticed small regression:
On **Assets** screen - the device status bar with white background while it was a blue one.

**iPhoneX view:**
![image uploaded from ios 11](https://user-images.githubusercontent.com/8749671/35978918-94e915f4-0cef-11e8-89e5-9953cc2d03d8.jpg)

**iPhone 5s view**:
![31412](https://user-images.githubusercontent.com/8749671/35978927-9aa72f76-0cef-11e8-882f-33a4b8d40b41.jpg)

**Develop and mockup view**:
<img width="651" alt="screen shot 2018-02-08 at 16 49 03" src="https://user-images.githubusercontent.com/8749671/35979076-09d4129c-0cf0-11e8-88d7-8d636442f906.png">


Could you please apply on 'Wallet' -> top-right button -> 'Manage Assets' -> 'Assets' screen the blue background to device status bar and fix conflicts?

flexsurfer commented:
@foopang could you please resolve conflicts and take a look on @Serhy comment, thanks

foopang commented:
@Serhy Thanks for checking! All layout issues should have been fixed now.

@flexsurfer @jeluard  something to note about the iPhone X screen issues, I did some hacks to workaround.

https://github.com/status-im/status-react/pull/2856/files#diff-d8aa2157cacb46da51afecaef8554e32R294 The bottom spacing is gone when in modal view. So I put an invisible `view` element to the bottom for adding bottom spacing. but it is not needed when the keyboard is opened up for signing transaction.

https://github.com/status-im/status-react/pull/2856/files#diff-93d9c10db297821b5145f29bdbb7b6bcR189 The transaction sent modal won't show but will block the layout from being navigated away when closing the transaction modal and immediately opening a new modal, so adding a delay here to fix.

https://github.com/status-im/status-react/pull/2856/files#diff-dc0057ed633d7dc9f584171ce43774d5R28 The top and bottom spacing has to be reduced from the height in order to get the screen viewport height of iPhone X so that "Scan QR code" screen can show properly.

https://github.com/status-im/status-react/pull/2856/files#diff-d8aa2157cacb46da51afecaef8554e32R275 The top bar and the bottom bar can't be two different background colors. To work around that I added a view with a different background color and stuck to the bottom so that it can show different background colors on Wallet screens.


<blockquote><img src="https://avatars1.githubusercontent.com/u/4666757?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>
<blockquote><img src="https://avatars1.githubusercontent.com/u/4666757?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>
<blockquote><img src="https://avatars1.githubusercontent.com/u/4666757?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>
<blockquote><img src="https://avatars1.githubusercontent.com/u/4666757?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-react/pull/2856">[Fix #2719] App does not fit on iPhone X screen by foopang · Pull Request #2856 · status-im/status-react</a></strong></div><div>fixes #2719
I guess we will need another splash image for iPhone X portrait (1125px x 2436px) 🙂


status: ready</div></blockquote>

Serhy commented:
Builds Android: https://i.diawi.com/jH87hH and iOS: https://i.diawi.com/ND1hGn has been tested on iPhoneX and Android 7.0 (Samsung J7).

It looks good, no issues and app fits on iPhoneX screen.

@foopang great job! Thank you a lot for the efforts in this PR!

Ready for merge!

chadyj commented:
Thank you @foopang! This has been a much requested feature and will be a fantastic enhancement to Status.
